### DISCUSSION TITLE
ECS, Archetypes, SPEED

### DISCUSSION THREAD

**Discussion by PJB3005** [space-wizards/RobustToolbox#3822] [Bikeshedding]
Our ECS is too slow, wish I had a time machine. How do we make it faster.

**Comment by PJB3005** [comment]
There's an existing library for archetype ECS in C# that seems to be gaining traction:
https://github.com/genaray/Arch
https://github.com/genaray/Arch.Extended

It may be worth at least taking as inspiration on some things.

**Reply by moonheart08** [reply]
Aside from the callback API this looks quite nice, stealing the way they handle systems would be good.

**Reply by metalgearsloth** [reply]
Is it worthwhile investigating just using it as a library to create less work then writing it ourselves and taking what's needed?

If it's viable we'd also get another ECS dev assisting the project (and they get more visibility of their project so win-win).

**Reply by moonheart08** [reply]
Personally think we'd be best off reusing an existing lib, recreating such a lib is a multi-year effort which we've already partially gone through and really shouldn't try to go through again.
Naturally we should **vendor it and have our own fork of the project**, as just dropping it in would be a bad idea, but beyond that Arch + Arch Extended seems like a viable way forward.

I do think that the ECS's "outer parallelism" (simultaneous dispatch of multiple systems at once) will need either disabled or major work to integrate into eventbus.

**Reply by metalgearsloth** [reply]
This will be real on god.

**Comment by PJB3005** [comment]
@DrSmugleaf and @Jezithyr already did a bit of work messing with some benchmarks here: https://github.com/space-wizards/RobustToolbox/blob/de00d2aca28824185090a2dc363d02dbade55b68/Robust.Benchmarks/EntityManager/ArchetypeComponentAccessBenchmark.cs

Intent was to make it kinda like Flecs, as I understand it: https://www.flecs.dev/flecs/

Using flecs itself is never going to be on the table due to C# memory management, but it's probably a good base.

**Reply by benev0** [reply]
I am somewhat confused why C# memory management would preclude flecs. My understanding is that as long as the structs that are placed under the flecs management do not contain pointers/objects no issues should arise. What am I missing?

**Reply by PJB3005** [reply]
Well, you're correct. Just good luck writing meaningful content code with only `unmanaged` type C# structs. That means no strings, lists, or anything fancy.

Also it's not 100% clear to me that flecs could be made babyproof such that content can't do nefarious shit with it.

**Reply by benev0** [reply]
An arc could be updated with component lifetime events to act as an interface between flecs and c# managed memory. This would burn some of the advantage gained by archetypes by requiring a lookup for objects. 

security is not my area of expertise, but the notion of adding an area outside of the c# sandbox would seem untenable under current assumptions.

**Comment by mirrorcult** [comment]
current consensus on this is we will arch ECS but component references need to be outright killed first to make it doable

**Reply by PJB3005** [reply]
Did anybody come up for a replacement for component references yet?

**Reply by mirrorcult** [reply]
put everything in shared

**Reply by mirrorcult** [reply]
besides that no
